plugins/woocommerce/includes/wc-rest-functions.php

```diff
@@ -424,5 +422,15 @@ function wc_rest_should_load_namespace( string $ns, string $rest_route = '' ): b
 		return true;
 	}
 
-	return str_starts_with( $rest_route, $ns );
+	/**
+	 * Filters whether a namespace should be loaded.
+	 *
+	 * @param bool   $should_load True if the namespace should be loaded, false otherwise.
+	 * @param string $ns          The namespace to check.
+	 * @param string $rest_route  The REST route being checked.
+	 * @param array  $known_namespaces Known namespaces that we know are safe to not load if the request is not for them.
+	 *
+	 * @since 9.2.3
+	 */
+	return apply_filters( 'wc_rest_should_load_namespace', str_starts_with( $rest_route, $ns ), $ns, $rest_route, $known_namespaces );
```

Comment:
Although I think this filter is necessary it will require all plugins that call an api internally to hook to it during setup to make sure the apis are loaded.

Is there a way for us to check if we're serving an internal request and automatically load the necessary namespace instead?

Internal requests can be done by calling `rest_preload_api_request1` or the following code for example.

```
$request = new WP_REST_Request( 'GET', 'endpoint' );
...
return rest_do_request( $request )->get_data();
```

One way I can think of would be to hook into [rest_pre_dispatch](https://developer.wordpress.org/reference/hooks/rest_pre_dispatch/) filter to check if it's going to be an internal request.

What do you think?

Reply:
@leonardola Seems like `rest_preload_api_request` should and does work already even without the API request context. For example when trying out this snippet:

```
add_action( 'wp_footer', function () {
	$memo = rest_preload_api_request([], '/wc/store/v1/cart' );
	echo "<pre>" . var_dump( $memo) . "</pre>";
});
```

It would correctly print the API response on a page where the API otherwise is not loaded. There are cases where this filter would still be needed, let me elaborate:

This issue happens when we are on a separate rest endpoint in one of Woo core's namespace, which then calls another endpoint that is also part of Woo core's namespace and therefore was not loaded when the `rest_api_init` was called. This means that the namespaces [listed here](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/wc-rest-functions.php#L402-L411) can't be making another internal API request to a different namespace (see this [comment](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/wc-rest-functions.php#L400)).

We expect, plugins to have a different namespace for their API request (as WP [recommends](https://github.com/WordPress/WordPress/blob/master/wp-includes/rest-api.php#L26)), and if this is followed then we will load all the namespaces allowing internal API requests. 

The only remaining case is when an existing namespace is used, which is also part of the above list, then only internal requests to the same namespace are allowed. So I am expecting that there won't be a widespread issue, but let me know what you think?

Reply:
@leonardola any further thoughts here?

Reply:
Nothing else to comment. I tested a custom endpoint and it loads all namespaces as expected. So just adding the filter should be enough to give flexibility to plugins that don't follow the recommendation to create a custom endpoint instead of using the woo namespace. 

I'm not sure why WooPayments does that but I'll check to prevent future problems.